### PR TITLE
feat: v1

### LINKED ISSUE #2: feat: suggestion resolution dates
When a moderator approves/considers/denies a suggestion, we have no way of knowing when that update happened. 

- One solution for this is to make a formatted timestamp based on the guild's locale (since timestamp syntax isn't supported in embed fields), which would be `DD/MM/YYYY` or `MM/DD/YYYY`. Rarely `YYYY/MM/DD` and even more rarely, `YYYY/DD/MM`.
- We can inject a formatted timestamp into the resolution's content. Much easier than the previous approach.

### LINKED ISSUE #3: refactor: simplification of suggestions formatting
Instead of adding a title being `Suggestion #{suggestion.id}`, we can move the ID to the footer and also add a timestamp with it. We can make the embed even cozier by moving author to footer.

### LINKED ISSUE #4: refactor: button-based modals for updating suggestions
Iris should always include a button for the author or moderators to update, clicking this button will open a modal that allows the user to update the contents.

The button will be automatically removed once the suggestion is accepted, considered, or denied.

### LINKED ISSUE #5: refactor: button-based modals for updating a suggestion's status
Locked to moderators, clicking this button will open a modal which will allow them to update a suggestion's status, being accepting, considering, or denying. Re-submitting an update will override the previous response and status, as opposed to stacking.

### LINKED ISSUE #7: feat: auto-threads
I was thinking of a few options:

- **None** (default): Iris creates the suggestion message, but does not provide any functionality to create threads, they will have to be manually created by the server's moderators. Useful for avoiding the creation of threads for spam/duplicate suggestions.
- ~~**Button**: Iris adds a new button to create a thread, however, it cannot be reactive, since she does not get gateway events for when the thread has already been created.~~
- **Automatic**: Always creates a thread for the suggested message.

### LINKED ISSUE #8: feat: auto-remove reactions on resolution
`#260` in `#suggestions`

### LINKED ISSUE #9: feat: auto-resolve suggestion references to message urls
For example, when somebody writes `Duplicated of #230`, Iris should replace it with `Duplicated of [#230](https://discord.com/channels/254360814063058944/257561807500214273/869738324137488416)`. This makes cross-referencing suggestions a lot easier.

### PR DESCRIPTION
**Pull request by kyranet**
Depends on:

- ~~https://github.com/skyra-project/archid-components/pull/74~~
- ~~https://github.com/skyra-project/archid-components/pull/75~~
- ~~https://github.com/skyra-project/archid-components/pull/76~~
- ~~https://github.com/skyra-project/archid-components/pull/77~~
- ~~https://github.com/skyra-project/archid-components/pull/78~~
- ~~https://github.com/skyra-project/archid-components/pull/79~~
- ~~https://github.com/skyra-project/archid-components/pull/80~~
- https://github.com/skyra-project/archid-components/pull/81
- https://github.com/skyra-project/archid-components/pull/82

### CODE REVIEW

File: crowdin.yml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+project_id: '512984'
```
Comment (favna) [inline crowdin.yml]:
```suggestion
project_id: ''
```

Empty for now until we know the actual ID after setting up the project

File: package.json [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+		"dev": "yarn build && yarn start",
+		"watch": "tsc -b src -w",
+		"watch:start": "tsc-watch -b src --onSuccess \"yarn start\"",
+		"clean": "node scripts/build/clean.mjs",
+		"start": "node --enable-source-maps dist/main.js",
+		"test": "eslint --ext ts src",
+		"lint": "eslint --fix --ext ts src",
+		"format": "prettier --write --loglevel=warn \"src/**/*.{js,ts,json}\"",
+		"update": "yarn upgrade-interactive",
+		"prepare": "husky install .github/husky"
```
Comment (favna) [inline package.json]:
```suggestion
		"postinstall": "husky install .github/husky"
```

prepare no worky for yarn v3 :( 
(yes other repos also need changing, feel free to do it if you feel like it. For bots postinstall works, for libraries see i.e. shapeshift where it uses pinst)

File: package.json [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+		}
+	},
+	"prettier": "@sapphire/prettier-config",
+	"eslintConfig": {
+		"extends": "@sapphire"
+	},
+	"packageManager": "yarn@3.2.1",
+	"volta": {
+		"node": "18.0.0"
+	}
```
Comment (favna) [inline package.json]: yay you copied this over

File: src/commands/suggest.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+export class UserCommand extends Command {
+	public override chatInputRun(interaction: Command.Interaction, options: Options): Command.AsyncResponse {
+		return options.id === undefined
+			? this.handleNew(interaction, options.suggestion)
+			: this.handleEdit(interaction, options.id, options.suggestion);
+	}
+
+	private async handleNew(_interaction: Command.Interaction, input: string) {
+		await Promise.resolve();
+		return this.message({ content: `TODO: ${input}` });
```
Comment (favna) [inline src/commands/suggest.ts]: 👀

File: src/main.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
+import { registerCommands } from '#lib/utilities/register-commands';
+import { envParseInteger, envParseString, setup } from '@skyra/env-utilities';
+import { Client, container } from '@skyra/http-framework';
+import { init, load } from '@skyra/http-framework-i18n';
+import { setInvite, setRepository } from '@skyra/shared-http-pieces';
+import '@skyra/shared-http-pieces/register';
```
Comment (favna) [inline src/main.ts:7]: This doesn't have to be at the top specifically?

**Review by favna — CHANGES_REQUESTED** [review]

File: src/commands/suggest.ts (hunk 4 above)
Comment (kyranet) [reply src/commands/suggest.ts]: 👀?

File: src/main.ts (hunk 5 above)
Comment (kyranet) [reply src/main.ts:7]: ~~No, but I can make it a dynamic import.~~ nvm I can't:

image

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+ENV LOG_LEVEL=info
+
+RUN apk add --no-cache dumb-init
+
+COPY --chown=node:node yarn.lock .
+COPY --chown=node:node package.json .
+COPY --chown=node:node .yarnrc.yml .
+COPY --chown=node:node .yarn/ .yarn/
+
+RUN sed -i 's/"prepare": "husky install .github\/husky"/"prepare": ""/' ./package.json
```
Comment (vladfrangu) [inline Dockerfile]: Why? You set HUSKY to false

File: Dockerfile [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+FROM base AS runner
+
+ENV NODE_ENV="production"
+ENV NODE_OPTIONS="--enable-source-maps"
+
+WORKDIR /usr/src/app
+
+COPY --chown=node:node --from=builder /usr/src/app/dist dist
+COPY --chown=node:node --from=builder /usr/src/app/src/locales src/locales
```
Comment (vladfrangu) [inline Dockerfile:48]: Isn't this supposed to be dist/locales? Or just `locales`?

File: src/lib/utilities/serialized-emoji.ts [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+	const index = emoji.lastIndexOf('.');
+	if (index === -1) throw new Error(`Invalid SerializedEmoji '${emoji}'`);
+
+	const name = emoji.slice(1, index);
+	const id = emoji.slice(index + 1);
+	return `${name}:${id}`;
+}
+
+const twemoji = /\p{Emoji_Presentation}/gu;
```
Comment (vladfrangu) [inline src/lib/utilities/serialized-emoji.ts]: not correct. Use the twemoji one / emoji-regex https://twemoji.maxcdn.com/v/latest/twemoji.js

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 7 above)
Comment (kyranet) [reply Dockerfile:48]: No, it expects the directory to be at `src/locales`, since they're not copied over.

File: package.json [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+	"eslintConfig": {
+		"extends": "@sapphire",
+		"rules": {
+			"@typescript-eslint/no-base-to-string": "off"
+		}
+	},
+	"packageManager": "yarn@3.2.1",
+	"volta": {
+		"node": "18.2.0",
+		"yarn": "1.22.18"
```
Comment (favna) [inline package.json]:
```suggestion
		"node": "18.2.0"
```

File: src/lib/common/types.ts [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+// eslint-disable-next-line @typescript-eslint/ban-types
+export type NonNullObject = {} & object;
+
+export function isUndefined(value: unknown): value is undefined {
+	return value === undefined;
+}
+
+export function isDefined<T>(value: T | undefined): value is T {
+	return value !== undefined;
+}
```
Comment (favna) [inline src/lib/common/types.ts]: Add to deps and use @sapphire/utilities, it's already in node_modules through hfx -> sapphire pieces anyway so it's not like it'll add overhead.

File: src/lib/utilities/serialized-emoji.ts [hunk 11]
```diff
@@ -0,0 +1,60 @@
+import type { IntegerString } from '#lib/common/types';
```
Comment (favna) [inline src/lib/utilities/serialized-emoji.ts:1]: Eventually we'll probably want to make @skyra/database-utilities in archid-components for stuff like this

**Review by favna — CHANGES_REQUESTED** [review]

File: package.json (hunk 9 above)
Comment (favna) [reply package.json]: bit of an oopsiewoopsie here

**Review by favna — APPROVED** [review]

**Pull request merged**